File: src/csharp/Foundgine.Providers/Storage/PostgresVector/PgVectorOptions.cs
namespace Foundgine.Providers.Storage.PostgresVector;

/// <summary>Distance function used to rank pgvector candidates.</summary>
public enum PgVectorDistance : byte
{
    /// <summary><c>&lt;=&gt;</c> — cosine distance. Typical default for text embeddings.</summary>
    Cosine,

    /// <summary><c>&lt;-&gt;</c> — Euclidean (L2) distance.</summary>
    L2,

    /// <summary><c>&lt;#&gt;</c> — negative inner product.</summary>
    InnerProduct
}

/// <summary>
/// Configuration for the pgvector-backed semantic lexicon table. This table
/// is a derived retrieval projection of the frozen semantic contract — it is
/// never the source of truth for schema topology.
/// </summary>
public sealed record PgVectorOptions(
    string TableName = "foundgine_semantic_lexicon",
    int Dimensions = 1536,
    PgVectorDistance Distance = PgVectorDistance.Cosine,
    string Schema = "public")
{
    public string QualifiedTableName => $"\"{Schema}\".\"{TableName}\"";
}

File: src/java/foundgine-providers/src/main/java/com/foundgine/providers/storage/postgresvector/PgVectorOptions.java
package com.foundgine.providers.storage.postgresvector;

/** Configuration for the pgvector-backed semantic lexicon projection. */
public record PgVectorOptions(String tableName, int dimensions, PgVectorDistance distance, String schema) {
	public PgVectorOptions() {
		this("foundgine_semantic_lexicon", 1536, PgVectorDistance.COSINE, "public");
	}

	public PgVectorOptions {
		if (tableName == null || tableName.isBlank())
			throw new IllegalArgumentException("tableName");
		if (dimensions < 1)
			throw new IllegalArgumentException("dimensions");
		if (distance == null)
			throw new NullPointerException("distance");
		if (schema == null || schema.isBlank())
			throw new IllegalArgumentException("schema");
	}

	public String qualifiedTableName() {
		return "\"" + quote(schema) + "\".\"" + quote(tableName) + "\"";
	}

	private static String quote(String value) {
		return value.replace("\"", "\"\"");
	}
}
